multi: fix incorrect multi_easy logic

jay opened the pull request:
- Use data->multi and not data->multi_easy to refer to the active multi.

Prior to this change there were a few places that incorrectly treated data->multi_easy, if != NULL, as the easy handle's active multi.

Background:

The easy handle's active multi is always data->multi. It points to either the user-created multi (if the multi interface is used) or the internally created multi (if the easy interface is used). The internally created multi is always data->multi_easy and is retained until the easy handle is cleaned up.

If an easy handle is used with the easy interface and then the multi interface, data->multi_easy and data->multi will both exist and be different.

Closes #xxxxx

---

There's one other that I didn't change because I'm not sure if the multi_easy should be checked before multi:

https://github.com/curl/curl/blob/8edcfedc1a144f438bd1cdf814a0016cbe678aaf/lib/connect.c#L270-L287

Some logic is used to get what should be the connection cache used by the easy handle. I notice `data->state.conn_cache` is not checked so I assume it's not available at that point.

bagder reviewed (approved):
:+1:

bagder commented:
@jay why not merge this?

jay commented:
> why not merge this?

I was unsure about Curl_getconnectinfo but I think I figured it out. Please review the squashme commit.

bagder commented on lib/multi.c:
```diff
@@ -898,7 +898,7 @@ CURLMcode curl_multi_remove_handle(struct Curl_multi *multi,
   Curl_detach_connection(data);
 
-  if(data->set.connect_only && !data->multi_easy) {
+  if(data->set.connect_only && (data->multi != data->multi_easy)) {
```
Why this change? If the easy interface is not used, the `multi_easy` pointer is NULL so isn't both the old and the new code achieving the same thing?

jay replied:
> If the easy interface is not used, the `multi_easy` pointer is NULL

sometimes. the rub is if an easy handle is used with the easy interface and then the multi interface. in that case they point to different multis. this pr addresses that issue by prioritizing multi over multi_easy  because the former is always the in-use multi. my hang up with Curl_getconnectinfo (which is used in this logic) is, following this scenario, what happens when the easy handle is removed by the user via curl_multi_remove_handle, because that would seem to cause disconnection of connections in multi_easy and not multi. afaict, this change fixes a bug by checking that they are different and Curl_getconnectinfo prioritizes multi over multi_easy

bagder commented on lib/connect.c:
```diff
@@ -282,9 +282,9 @@ curl_socket_t Curl_getconnectinfo(struct Curl_easy *data,
-                           &data->multi->conn_cache, &find, conn_is_conn);
+                           data->multi?
+                           &data->multi->conn_cache:
+                           &data->multi_easy->conn_cache, &find, conn_is_conn);
```
I mean, it can hardly reach this code with `data->multi` being NULL, can it ?

bagder commented:
@jay I propose  #12992 so that each easy handle only has one multi handle to relate to.

bagder commented:
Now merged. With this, an easy handle can only ever be associated with a single multi handle.

bagder reviewed (changes requested):
This can now be adjusted

bagder commented:
@jay do you want me to continue this?

I believe `Curl_close()` can also be simplified somewhat (when the easy handle is found to be part of a multi handle).

jay commented:
> do you want me to continue this?
> 
> I believe `Curl_close()` can also be simplified somewhat (when the easy handle is found to be part of a multi handle).

I dropped the fixup commit and changed the other commit's message, if more needs to be done have at it!

jay commented:
@bagder can you have another look, I don't think there's anything left to add

bagder reviewed (approved):
Looks good!
